src/fparser/Fortran2003.py

```diff
@@ -6751,63 +6905,108 @@ class Module_Subprogram(Base): # R1108
+        '''
+        :param str string: Fortran code to check for a match
+        :return: 5-tuple containing strings and instances of the classes
+                 describing a module (optional module nature, optional
+                 double colon delimiter, mandatory module name, optional
+                 "ONLY" specification and optional "Rename" or "Only" list)
+        :rtype: 5-tuple of objects (module name and 4 optional)
+        '''
+        if string[:3].upper() != 'USE':
+            return
         line = string[3:]
-        if not line: return
-        if isalnum(line[0]): return
+        if not line:
+            return
+        if isalnum(line[0]):
+            return
         line = line.lstrip()
         i = line.find('::')
         nature = None
-        if i!=-1:
+        dcolon = None
+        if i != -1:
+            dcolon = '::'
             if line.startswith(','):
-                l = line[1:i].strip()
-                if not l: return
-                nature = Module_Nature(l)
+                line_nat = line[1:i].strip()
+                if not line_nat:
+                    return
+                nature = Module_Nature(line_nat)
             line = line[i+2:].lstrip()
-            if not line: return
+            if not line:
+                return
         i = line.find(',')
-        if i==-1: return nature, Module_Name(line), '', None
+        if i == -1:
+            return nature, dcolon, Module_Name(line), '', None
         name = line[:i].rstrip()
-        if not name: return
+        if not name:
+            return
         name = Module_Name(name)
         line = line[i+1:].lstrip()
-        if not line: return
+        if not line:
+            return
         if line[:4].upper() == 'ONLY':
             line = line[4:].lstrip()
             if line[0] != ':':
                 return
             line = line[1:].lstrip()
             if not line:
-                return nature, name, ', ONLY:', None
-            return nature, name, ', ONLY:', Only_List(line)
-        return nature, name, ',', Rename_List(line)
-    match = staticmethod(match)
+                return nature, dcolon, name, ', ONLY:', None
+            return nature, dcolon, name, ', ONLY:', Only_List(line)
+        return nature, dcolon, name, ',', Rename_List(line)
+
     def tostr(self):
-        s = 'USE'
-        if self.items[0] is not None:
-            s += ', %s' % (self.items[0])
-        s += ' :: %s%s' % (self.items[1], self.items[2])
-        if self.items[3] is not None:
-            s += ' %s' % (self.items[3])
-        return s
+        '''
+        :return: parsed representation of "USE" statement
+        :rtype: string
+        '''
+        usestmt = 'USE'
+        if self.items[0] is not None and self.items[1] is not None:
+            usestmt += ', %s %s' % (self.items[0], self.items[1])
+        elif self.items[0] is not None and self.items[1] is None:
+            message = 'Module nature must be followed by "::"'
```

Comment:
Please add (to) a test to cover this case.

Reply:
It turned out that the check for missing `::` after the `Module_Nature` was not in the `match(string)` method at all so I added it. This now raises a `NoMatchError` before `tostr(self)` method is reached so I deleted the above lines from the `tostr(self)` method and added a test to check that the `NoMatchError` is raised correctly. 
I think all lines in `class Use_Stmt` are now covered by tests.